### ISSUE TITLE
ASI for `++` and `--` tokens

### ISSUE THREAD

**Issue by maciejhirsz** [ratel-rust/ratel-core#78]
```js
x 
++ 
y
```
and

```js
x 
-- 
y
```

Should be parsed as `x; ++y;` and `x; --y;` respectively.
They are currently parsed as `x++; y;` and `x--; y;`

**Comment by s-coimbra21** [comment]
there is also

```js
x++ + ++y
```

which the codegen prints as 
```js
x+++++y
```

then, if we try to parse that it becomes a binary expression with the left hand side being a double postfix increment but it should actually be a syntax error

parsing ```x++ + ++x``` works fine however but generates ```x+++++y``` which is wrong

**Comment by cmtt** [comment]
Hello @s-coimbra21, I cannot acknowledge the issue with the source 'x++ + ++y`.

The syntax is also valid JavaScript with the following AST:

````json
{
  "type": "Program",
  "body": [
    {
      "type": "ExpressionStatement",
      "expression": {
        "type": "BinaryExpression",
        "operator": "+",
        "left": {
          "type": "UpdateExpression",
          "operator": "++",
          "argument": {
            "type": "Identifier",
            "name": "x",
            "start": 0,
            "end": 1
          },
          "prefix": false,
          "start": 0,
          "end": 3
        },
        "right": {
          "type": "UpdateExpression",
          "operator": "++",
          "argument": {
            "type": "Identifier",
            "name": "y",
            "start": 8,
            "end": 9
          },
          "prefix": true,
          "start": 6,
          "end": 9
        },
        "start": 0,
        "end": 9
      },
      "start": 0,
      "end": 9
    }
  ],
  "start": 0,
  "end": 9
}
````
The currently deployed version on http://maciej.codes/ratel-wasm generates the same code again.
However, the returned ranges in source differ currently from master.

**Comment by s-coimbra21** [comment]
My apologies, the first snippet works as intended. The error is the code that is generated from that ```x+++++y``` which should be a syntax error but which the ratel parser accepts though it produces an invalid AST (basically it parses ```(x++)++ + y```)

**Comment by maciejhirsz** [comment]
Ye, there are two issues here, one is not checking validity of left value (`(x++)` is an invalid expression for `++` or any assignment operators). Another is that codegen should wrap the middle `+` in spaces.

**Comment by maciejhirsz** [comment]
Fixed with #85
